From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sun, 19 Jul 2026 12:39:25 -0500
Subject: [PATCH] Fish Handle Mid Tick Tasks

This patch re-enabled mid-tick tasks.
Instead of iterating through all worlds, if the requesting thread is a world thread, it will only
  execute its own mid-tick tasks.

diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index a186148dc817038788f79e4a74b18238692cd23b..40ef01f261cc61f1d9f458a7fa3e367e0ba17c47 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -459,6 +459,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @Override
     public final void moonrise$executeMidTickTasks() {
+        // Fish start - Handle Mid-Tick Tasks
+        if (me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled && Thread.currentThread() instanceof me.biquaternions.fish.async.thread.WorldTickThread worldThread) {
+            me.biquaternions.fish.async.AsyncWorldTicking.executeMidTickTasks(worldThread.getTickingWorld());
+            return;
+        }
+        // Fish end - Handle Mid-Tick Tasks
         final long startTime = System.nanoTime();
         if ((startTime - this.lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - this.lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) {
             // it's shown to be bad to constantly hit the queue (chunk loads slow to a crawl), even if no tasks are executed.
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 3019b16b00bee292322f0a83f3d7715eea25bbfe..36cad318ac0724420a75268741ae5aaa8ff49f9b 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -179,7 +179,6 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
             // call mid-tick tasks for chunk system
             if ((i & 7) == 0) {
-                if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) // SparklyPaper - parallel world ticking (only run mid-tick at the end of each tick / fixes concurrency bugs related to executeMidTickTasks)
                 ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks();
                 continue;
             }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 729569f750e942c790ed0946c517c454eafeacae..6120a2a6c32e7dd49be269f6a283959d44ad20ad 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1380,7 +1380,6 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) // SparklyPaper - parallel world ticking
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
         }
         // Paper end - rewrite chunk system
@@ -1394,7 +1393,6 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         }
         // Paper start - rewrite chunk system
         if ((++this.tickedBlocksOrFluids & 7L) != 0L) {
-            if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) // SparklyPaper - parallel world ticking
             ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.server).moonrise$executeMidTickTasks();
         }
         // Paper end - rewrite chunk system
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index b9eae69c4c177e12838ab879a17f51fb7c7a1a6c..76adc99ba4f9f05799fdb193e7b199a430fea0a4 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1509,7 +1509,6 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 if ((++tickedEntities & 7) == 0) {
-                    if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) // SparklyPaper - parallel world ticking
                     ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)(Level)(Object)this).moonrise$midTickTasks();
                 }
                 // Paper end - rewrite chunk system
@@ -1531,7 +1530,6 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
             // Paper end - Prevent block entity and entity crashes
         }
-        if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) // SparklyPaper - parallel world ticking
         this.moonrise$midTickTasks(); // Paper - rewrite chunk system
     }
 
